File: sdk/python/examples/custom_agent.py
"""
ARKVOID SDK – Custom Async AI Agent (Python)

Demonstrates tracing a multi-step agent with:
- AsyncArkvoidClient
- Session grouping
- Parent/child trace chains
- Tool call recording
- Risk escalation on failure

pip install arkvoid openai aiohttp
"""

import asyncio
import json
import os
import time
import uuid
from typing import Any, Dict, List, Optional

from openai import AsyncOpenAI
from arkvoid import AsyncArkvoidClient, ToolCallRecord

# ─────────────────────────────────────────────
# Setup
# ─────────────────────────────────────────────

openai_client = AsyncOpenAI(api_key=os.environ["OPENAI_API_KEY"])

TOOLS = [
    {
        "type": "function",
        "function": {
            "name": "web_search",
            "description": "Search the web for current information",
            "parameters": {
                "type": "object",
                "properties": {"query": {"type": "string"}},
                "required": ["query"],
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "calculate",
            "description": "Evaluate a mathematical expression",
            "parameters": {
                "type": "object",
                "properties": {"expression": {"type": "string"}},
                "required": ["expression"],
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "save_result",
            "description": "Persist a result to the database",
            "parameters": {
                "type": "object",
                "properties": {
                    "key": {"type": "string"},
                    "value": {"type": "string"},
                },
                "required": ["key", "value"],
            },
        },
    },
]


# ─────────────────────────────────────────────
# Simulated tool execution
# ─────────────────────────────────────────────

async def execute_tool(name: str, args: Dict[str, str]) -> Dict[str, Any]:
    await asyncio.sleep(0.1)  # simulate latency
    if name == "web_search":
        return {"results": [f"Result about: {args.get('query', '')}"], "count": 3}
    elif name == "calculate":
        try:
            return {"result": eval(args.get("expression", "0"), {"__builtins__": {}}, {})}
        except Exception as e:
            return {"error": str(e)}
    elif name == "save_result":
        return {"saved": True, "key": args.get("key"), "timestamp": time.time()}
    return {"error": "Unknown tool"}


# ─────────────────────────────────────────────
# Agent loop
# ─────────────────────────────────────────────

async def run_agent(
    task: str,
    user_id: str,
    arkvoid: AsyncArkvoidClient,
    max_iterations: int = 6,
) -> Dict[str, Any]:
    session_id = str(uuid.uuid4())
    model = "gpt-4o"

    print(f"\n🤖 Session: {session_id}")
    print(f"📋 Task: {task}\n")

    messages: List[Dict[str, Any]] = [
        {"role": "system", "content": "You are a capable research assistant. Use tools to answer thoroughly."},
        {"role": "user", "content": task},
    ]

    parent_trace_id: Optional[str] = None
    iteration = 0
    final_answer: Optional[str] = None

    for iteration in range(1, max_iterations + 1):
        iter_start = time.time()
        tool_calls_this_iter: List[ToolCallRecord] = []

        print(f"  → Iteration {iteration}")

        response = await openai_client.chat.completions.create(
            model=model,
            messages=messages,
            tools=TOOLS,
        )
        message = response.choices[0].message
        messages.append(message.model_dump(exclude_unset=True))

        # Agent finished
        if not message.tool_calls:
            final_answer = message.content or ""
            trace = await arkvoid.trace(
                action="agent_final_answer",
                risk_level="low",
                session_id=session_id,
                parent_trace_id=parent_trace_id,
                model_provider="openai",
                model_name=model,
                input_tokens=response.usage.prompt_tokens if response.usage else None,
                output_tokens=response.usage.completion_tokens if response.usage else None,
                duration_ms=int((time.time() - iter_start) * 1000),
                output_data=final_answer,
                metadata={"user_id": user_id, "iteration": iteration},
                tags=["agent", "final"],
            )
            print(f"  ✅ Done → trace: {trace.trace_id if trace else '–'}")
            break

        # Execute tool calls
        for tc in message.tool_calls:
            tool_name = tc.function.name
            tool_args = json.loads(tc.function.arguments)
            tool_start = time.time()

            risk_level = "medium" if tool_name == "save_result" else "low"
            result = await execute_tool(tool_name, tool_args)
            tool_latency = int((time.time() - tool_start) * 1000)

            tool_calls_this_iter.append(
                ToolCallRecord(
                    tool_name=tool_name,
                    call_index=len(tool_calls_this_iter),
                    input=tool_args,
                    output=result,
                    latency_ms=tool_latency,
                    status="error" if "error" in result else "success",
                )
            )

            messages.append({
                "role": "tool",
                "tool_call_id": tc.id,
                "content": json.dumps(result),
            })

            print(f"     🔧 {tool_name} → {str(result)[:60]}")

        # Trace this iteration
        trace = await arkvoid.trace(
            action=f"agent_iteration_{iteration}",
            risk_level="low",
            session_id=session_id,
            parent_trace_id=parent_trace_id,
            model_provider="openai",
            model_name=model,
            input_tokens=response.usage.prompt_tokens if response.usage else None,
            output_tokens=response.usage.completion_tokens if response.usage else None,
            duration_ms=int((time.time() - iter_start) * 1000),
            tool_calls=tool_calls_this_iter,
            metadata={
                "user_id": user_id,
                "iteration": iteration,
                "tool_count": len(tool_calls_this_iter),
            },
            tags=["agent", "iteration"],
        )

        if not parent_trace_id and trace:
            parent_trace_id = trace.trace_id

    return {
        "session_id": session_id,
        "final_answer": final_answer,
        "iterations": iteration,
        "parent_trace_id": parent_trace_id,
    }


# ─────────────────────────────────────────────
# Run
# ─────────────────────────────────────────────

async def main():
    print("🤖 ARKVOID Custom Async Agent (Python)\n")

    async with AsyncArkvoidClient(
        api_key=os.environ["ARKVOID_API_KEY"],
        agent="research-agent",
        debug=True,
    ) as arkvoid:
        result = await run_agent(
            task="What is 25% of 12,400? Also find recent news about AI regulation.",
            user_id="user_demo_001",
            arkvoid=arkvoid,
        )

    print(f"\n────────────────────────────────────")
    print(f"Session:    {result['session_id']}")
    print(f"Iterations: {result['iterations']}")
    print(f"Root trace: {result['parent_trace_id']}")
    if result["final_answer"]:
        print(f"Answer:     {result['final_answer'][:200]}...")


if __name__ == "__main__":
    asyncio.run(main())

File: sdk/javascript/examples/custom-agent.ts
/**
 * ARKVOID SDK – Custom AI Agent Tracing Example
 *
 * Demonstrates tracing a multi-step AI agent with:
 * - Session-level grouping
 * - Parent/child trace relationships
 * - Tool call tracking
 * - State delta recording
 * - Risk escalation
 *
 * npm install arkvoid openai
 */

import { randomUUID } from "node:crypto";
import OpenAI from "openai";
import { ArkvoidClient } from "arkvoid";
import type { ToolCallRecord, RiskLevel } from "arkvoid";

// ─────────────────────────────────────────────
// Setup
// ─────────────────────────────────────────────

const openai = new OpenAI({ apiKey: process.env.OPENAI_API_KEY });

const arkvoid = new ArkvoidClient({
  apiKey: process.env.ARKVOID_API_KEY!,
  agent: "research-agent",
  environment: "production",
  debug: true,
});

// ─────────────────────────────────────────────
// Tool Definitions
// ─────────────────────────────────────────────

const TOOLS: OpenAI.Chat.ChatCompletionTool[] = [
  {
    type: "function",
    function: {
      name: "web_search",
      description: "Search the web for current information",
      parameters: {
        type: "object",
        properties: {
          query: { type: "string", description: "Search query" },
        },
        required: ["query"],
      },
    },
  },
  {
    type: "function",
    function: {
      name: "calculate",
      description: "Perform mathematical calculations",
      parameters: {
        type: "object",
        properties: {
          expression: {
            type: "string",
            description: "Mathematical expression to evaluate",
          },
        },
        required: ["expression"],
      },
    },
  },
  {
    type: "function",
    function: {
      name: "save_result",
      description: "Save a result to the database",
      parameters: {
        type: "object",
        properties: {
          key: { type: "string" },
          value: { type: "string" },
        },
        required: ["key", "value"],
      },
    },
  },
];

// ─────────────────────────────────────────────
// Simulated Tool Execution
// ─────────────────────────────────────────────

async function executeTool(
  name: string,
  args: Record<string, string>
): Promise<{ result: string; latencyMs: number }> {
  const start = Date.now();

  // Simulate tool execution
  await new Promise((r) => setTimeout(r, Math.random() * 300 + 100));

  let result: string;
  switch (name) {
    case "web_search":
      result = `Search results for "${args.query}": [Simulated results about ${args.query}]`;
      break;
    case "calculate":
      result = String(eval(args.expression ?? "0")); // demo only!
      break;
    case "save_result":
      result = `Saved: ${args.key} = ${args.value}`;
      break;
    default:
      result = "Unknown tool";
  }

  return { result, latencyMs: Date.now() - start };
}

// ─────────────────────────────────────────────
// Agent Loop with Full Tracing
// ─────────────────────────────────────────────

interface AgentRunOptions {
  task: string;
  userId: string;
  maxIterations?: number;
}

async function runAgent({ task, userId, maxIterations = 5 }: AgentRunOptions) {
  const sessionId = randomUUID();
  const model = "gpt-4o";

  console.log(`\n🤖 Agent session: ${sessionId}`);
  console.log(`📋 Task: ${task}\n`);

  const messages: OpenAI.Chat.ChatCompletionMessageParam[] = [
    {
      role: "system",
      content:
        "You are a research assistant. Use tools to answer questions thoroughly.",
    },
    { role: "user", content: task },
  ];

  let iteration = 0;
  let parentTraceId: string | undefined;
  let finalAnswer: string | undefined;

  while (iteration < maxIterations) {
    iteration++;
    const iterStart = Date.now();
    const toolCallsThisIter: ToolCallRecord[] = [];

    console.log(`\n→ Iteration ${iteration}`);

    const response = await openai.chat.completions.create({
      model,
      messages,
      tools: TOOLS,
    });

    const message = response.choices[0]?.message;
    if (!message) break;

    messages.push(message);

    // No more tool calls → agent is done
    if (!message.tool_calls?.length) {
      finalAnswer = message.content ?? "";

      // Trace final answer
      const traceResult = await arkvoid.trace({
        action: "agent_final_answer",
        riskLevel: "low",
        sessionId,
        parentTraceId,
        modelProvider: "openai",
        modelName: model,
        inputTokens: response.usage?.prompt_tokens,
        outputTokens: response.usage?.completion_tokens,
        durationMs: Date.now() - iterStart,
        outputData: finalAnswer,
        metadata: {
          userId,
          iteration,
          taskSummary: task.slice(0, 100),
        },
        tags: ["agent", "final-answer"],
      });

      console.log(`✅ Final trace: ${traceResult?.traceId}`);
      break;
    }

    // Execute tool calls
    for (const toolCall of message.tool_calls) {
      const toolName = toolCall.function.name;
      const toolArgs = JSON.parse(toolCall.function.arguments) as Record<
        string,
        string
      >;

      const riskLevel: RiskLevel =
        toolName === "save_result"
          ? "medium"
          : toolName === "web_search"
            ? "low"
            : "low";

      const { result, latencyMs } = await executeTool(toolName, toolArgs);

      toolCallsThisIter.push({
        toolName,
        callIndex: toolCallsThisIter.length,
        input: toolArgs,
        output: { result },
        latencyMs,
        status: "success",
      });

      console.log(`  🔧 ${toolName}(${JSON.stringify(toolArgs)}) → ${result.slice(0, 60)}`);

      messages.push({
        role: "tool",
        tool_call_id: toolCall.id,
        content: result,
      });
    }

    // Trace this iteration
    const traceResult = await arkvoid.trace({
      action: `agent_iteration_${iteration}`,
      riskLevel: "low",
      sessionId,
      parentTraceId,
      modelProvider: "openai",
      modelName: model,
      inputTokens: response.usage?.prompt_tokens,
      outputTokens: response.usage?.completion_tokens,
      durationMs: Date.now() - iterStart,
      toolCalls: toolCallsThisIter,
      metadata: {
        userId,
        iteration,
        toolCount: toolCallsThisIter.length,
        tools: toolCallsThisIter.map((t) => t.toolName),
      },
      tags: ["agent", "iteration"],
    });

    if (!parentTraceId && traceResult?.traceId) {
      parentTraceId = traceResult.traceId;
    }
  }

  return { sessionId, finalAnswer, iterations: iteration };
}

// ─────────────────────────────────────────────
// Run
// ─────────────────────────────────────────────

(async () => {
  console.log("🤖 ARKVOID Custom Agent Tracing Example\n");

  const { sessionId, finalAnswer, iterations } = await runAgent({
    task: "What is 15% of 2847, and search for recent AI governance news?",
    userId: "user_demo_001",
  });

  console.log(`\n─────────────────────────────`);
  console.log(`Session: ${sessionId}`);
  console.log(`Iterations: ${iterations}`);
  console.log(`Answer: ${finalAnswer?.slice(0, 200)}...`);
})();
